### PR TITLE
New rule S6640: Allowing unsafe code is security-sensitive

### LINKED ISSUE #7290: New rule S6640: Allowing unsafe code is security-sensitive
## Why
C# allows for unsafe code blocks, where it is possible to define pointer types, fixed buffers, to use manual memory allocation,... The CLR cannot verify the safety of this code, therefore it can in some circumstances lead to unexpected behavior and security risks.

## Detection logic
All unsafe code blocks should be detected. This can either be a keyword in a method header, or a keyword for a block expression. Examples and more information can be found in the [language reference](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/unsafe).

#### Sensitive
```csharp
private unsafe int subarraySum(int[] array, int start, int end)  // Sensitive
{
    var sum = 0;

    // Skip array bound checks for extra performance
    fixed (int* firstNumber = array)
    {
        for (int i = start; i < end; i++)
            sum += *(firstNumber + i);
    }

    return sum;
}
```

#### Compliant
```csharp
private int subarraySum(int[] array, int start, int end)
{
    var sum = 0;

    Span<int> span = array.AsSpan();
    for (int i = start; i < end; i++)
        sum += span[i];

    return sum;
}
```

## RSPEC
The relevant RSPEC PR can be found [here](https://github.com/SonarSource/rspec/pull/1945). It also contains additional information regarding the issue message and highlighting.

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixes #7290

### CODE REVIEW

File: analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S6640.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Remote\Serialization\LruBoundedCache.cs",
+"region":  {
+"startLine":  61,
+"startColumn":  13,
+"endLine":  61,
+"endColumn":  19
+}
+}
+}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S6640.json:15]: TP

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxTokenListExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Extensions
+{
+    public static class SyntaxTokenListExtensions
+    {
+        public static SyntaxToken? Find(this SyntaxTokenList tokenList, SyntaxKind kind) =>
+            tokenList.IndexOf(kind) is var index && index != -1
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxTokenListExtensions.cs:26]: Alternative (I'm not sure whether it is better, so the change is optional):
```suggestion
            tokenList.IndexOf(kind) is >= 0 and var index
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/UnsafeCodeBlocks.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    void LocalFunction()
+    {
+        unsafe void Local(byte* pointer) { }                        // FN
+        void SafeLocal(byte noPointer) { }                          // Compliant
+    }
+
+    unsafe class UnsafeClass { }                                    // FN
+
+    unsafe struct UnsafeStruct                                      // FN
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/UnsafeCodeBlocks.cs:26]: `record` and `record struct` are missing.

File: analyzers/rspec/cs/S6640.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  "remediation": {
+    "func": "Constant\/Issue",
+    "constantCost": "60min"
+  },
+  "tags": [],
+  "defaultSeverity": "Major",
+  "ruleSpecification": "RSPEC-6640",
+  "sqKey": "S6640",
+  "scope": "All",
+  "quickfix": "unknown"
```
Comment (martin-strecker-sonarsource) [inline analyzers/rspec/cs/S6640.json:14]: Should be infeasible.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
RSpec needs an update for the quickfix field.

File: analyzers/rspec/cs/S6640.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  "remediation": {
+    "func": "Constant\/Issue",
+    "constantCost": "60min"
+  },
+  "tags": [],
+  "defaultSeverity": "Major",
+  "ruleSpecification": "RSPEC-6640",
+  "sqKey": "S6640",
+  "scope": "All",
+  "quickfix": "infeasible"
```
Comment (martin-strecker-sonarsource) [inline analyzers/rspec/cs/S6640.json:14]: Your RSpec change was merged by me. You may want to run the update script again against RSpec master, just to make sure everything is as expected

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. Make sure the RSpec is up-to-date with master as the RSpec change PR is merged now.

**Pull request merged**